### PR TITLE
refactor: logging in many controller

### PR DESCRIPTION
**Pull request by wy-lucky**
Refactors the federatedcluster, follower, policyrc, statusaggregator controller to follow the new logging conventions

### CODE REVIEW

File: pkg/controllers/statusaggregator/controller.go [hunk 1]
```diff
@@ -243,47 +249,48 @@ func (a *StatusAggregator) HasSynced() bool {
...
 		a.metrics.Duration("status-aggregator.latency", startTime)
-		klog.V(4).
-			Infof("status aggregator for %v finished reconciling %v (duration: %v)", sourceKind, key, time.Since(startTime))
+		logger.V(3).Info("status aggregator finished reconciling")
 	}()
 
 	sourceObject, err := objectFromCache(a.sourceStore, key)
 	if err != nil {
-		utilruntime.HandleError(err)
+		logger.Error(err, "")
```
Comment (qclc) [inline pkg/controllers/statusaggregator/controller.go:265]: How about adding an error log here, for example: 
```suggestion
		logger.Error(err, "Failed to get source object from cache")
```

File: pkg/controllers/statusaggregator/controller.go [hunk 2]
```diff
@@ -353,7 +362,7 @@ func (a *StatusAggregator) clusterObjs(qualifiedName common.QualifiedName) (map[
 		)
 		if err != nil {
 			wrappedErr := errors.Wrapf(err, "Failed to get %s %q from cluster %q", targetKind, key, cluster.Name)
-			utilruntime.HandleError(wrappedErr)
+			logger.Error(wrappedErr, "")
```
Comment (qclc) [inline pkg/controllers/statusaggregator/controller.go:365]: How about adding an error log here？
```suggestion
			logger.WithValues("target-kind", targetKind, "object", key, "cluster-name", cluster.Name).Error(err, "Failed to get object from cluster")
```

Comment (qclc) [reply pkg/controllers/statusaggregator/controller.go:365]: Please add other error messages if missing.

Comment (wy-lucky) [reply pkg/controllers/statusaggregator/controller.go]: get

File: pkg/controllers/statusaggregator/controller.go [hunk 3]
```diff
@@ -192,11 +195,14 @@ func newStatusAggregator(controllerConfig *util.ControllerConfig,
 	if err != nil {
 		return nil, err
 	}
-	klog.Infof("Status aggregator %q NewResourceInformer", federatedAPIResource.Kind)
+	a.logger.WithValues("federated-api-resource-kind", federatedAPIResource.Kind).Info("Status aggregator NewResourceInformer")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/controller.go:198]: We can remove this log

File: pkg/controllers/statusaggregator/controller.go [hunk 4]
```diff
@@ -131,6 +133,7 @@ func newStatusAggregator(controllerConfig *util.ControllerConfig,
 		typeConfig:    typeConfig,
 		plugin:        plugin,
 		metrics:       controllerConfig.Metrics,
+		logger:        klog.LoggerWithName(klog.Background(), ControllerName),
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/controller.go:136]: The convention is not to use a named logger, but to inject the controller name as a key-value. `klog.LoggerWithValues(klog.Background(), "controller", ControllerName, "ftc", typeConfig.Name)`

File: pkg/controllers/statusaggregator/controller.go [hunk 5]
```diff
@@ -206,15 +212,15 @@ func (a *StatusAggregator) Run(stopChan <-chan struct{}) {
 	go a.clusterDeliverer.RunMetricLoop(stopChan, 30*time.Second, a.metrics,
 		delayingdeliver.NewMetricTags("schedulingpreference-clusterDeliverer", a.typeConfig.GetTargetType().Kind))
 	if !cache.WaitForNamedCacheSync(a.name, stopChan, a.HasSynced) {
-		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync for controller: %s", a.name))
+		a.logger.WithValues("name", a.name).Error(errors.New(""), "timed out waiting for caches to sync for controller")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/controller.go:215]:
```suggestion
		return
```

Comment (limhawjia) [reply pkg/controllers/statusaggregator/controller.go:215]: `cache.WaitForNamedCacheSync` already logs when it times out

File: pkg/controllers/statusaggregator/controller.go [hunk 6]
```diff
@@ -206,15 +212,15 @@ func (a *StatusAggregator) Run(stopChan <-chan struct{}) {
...
 	}
 	a.worker.Run(stopChan)
 
 	// Ensure all goroutines are cleaned up when the stop channel closes
 	go func() {
 		defer func() {
 			if r := recover(); r != nil {
-				klog.Warningf("recovered from panic: %v", r)
+				a.logger.Error(errors.New("r"), "recovered from panic")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/controller.go:223]:
```suggestion
				a.logger.Error(fmt.Errorf("%v",  r), "recovered from panic")
```

Comment (limhawjia) [reply pkg/controllers/statusaggregator/controller.go:223]: To be honest, the `recover` here seems a little strange and unnecessary... But let's just focus on logging for this PR

File: pkg/controllers/statusaggregator/controller.go [hunk 7]
```diff
@@ -225,15 +231,15 @@ func (a *StatusAggregator) Run(stopChan <-chan struct{}) {
 
 func (a *StatusAggregator) HasSynced() bool {
 	if !a.informer.ClustersSynced() {
-		klog.V(2).Infof("Cluster list not synced")
+		a.logger.V(2).Info("Cluster list not synced")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/controller.go:234]:
```suggestion
		a.logger.V(3).Info("Cluster list not synced")
```

File: pkg/controllers/statusaggregator/controller.go [hunk 8]
```diff
@@ -225,15 +231,15 @@ func (a *StatusAggregator) Run(stopChan <-chan struct{}) {
...
 		return false
 	}
 	if !a.federatedController.HasSynced() {
-		klog.V(2).Infof("Federated type not synced")
+		a.logger.V(2).Info("Federated type not synced")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/controller.go:238]:
```suggestion
		a.logger.V(3).Info("Federated type not synced")
```

File: pkg/controllers/statusaggregator/controller.go [hunk 9]
```diff
@@ -225,15 +231,15 @@ func (a *StatusAggregator) Run(stopChan <-chan struct{}) {
...
 		return false
 	}
 	if !a.sourceController.HasSynced() {
-		klog.V(2).Infof("Status not synced")
+		a.logger.V(2).Info("Status not synced")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/controller.go:242]:
```suggestion
		a.logger.V(3).Info("Status not synced")
```

File: pkg/controllers/statusaggregator/controller.go [hunk 10]
```diff
@@ -294,7 +301,7 @@ func (a *StatusAggregator) reconcile(qualifiedName common.QualifiedName) worker.
 	}
 
 	if needUpdate {
-		klog.V(4).Infof("about to update status of source object %v: %v", sourceKind, key)
+		logger.V(1).Info("About to update status of source object")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/controller.go:304]:
```suggestion
		logger.V(1).Info("Updating status of source object")
```

File: pkg/controllers/statusaggregator/controller.go [hunk 11]
```diff
@@ -314,7 +321,7 @@ func (a *StatusAggregator) reconcile(qualifiedName common.QualifiedName) worker.
 		sourcefeedback.PopulateStatusAnnotation(newObj, clusterObjs, &needUpdate)
 
 		if needUpdate {
-			klog.V(4).Infof("about to update annotation of source object %v: %v", sourceKind, key)
+			logger.V(1).Info("About to update annotation of source object")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/controller.go:324]:
```suggestion
			logger.V(1).Info("Updating annotation of source object")
```

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 12]
```diff
@@ -39,9 +39,12 @@ func NewDeploymentPlugin() *DeploymentPlugin {
 }
 
 func (receiver *DeploymentPlugin) AggregateStatues(
+	ctx context.Context,
 	sourceObject, fedObject *unstructured.Unstructured,
 	clusterObjs map[string]interface{},
 ) (*unstructured.Unstructured, bool, error) {
+	logger := klog.FromContext(ctx).WithValues("source-object-name", sourceObject.GetName())
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/deployment.go:46]:
```suggestion
	logger := klog.FromContext(ctx).WithValues("source-object-name", sourceObject.GetName(), "status-aggregator-plugin", "deployments")
```

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 13]
```diff
@@ -60,11 +63,8 @@ func (receiver *DeploymentPlugin) AggregateStatues(
 		var found bool
 		status, found, err := unstructured.NestedMap(utd.Object, common.StatusField)
 		if err != nil || !found {
-			klog.Errorf(
-				"Failed to get status of cluster resource object deployment %q for cluster %q",
-				sourceObject.GetName(),
-				clusterName,
-			)
+			logger.WithValues("cluster-name", clusterName).
+				Error(err, "Failed to get status of cluster resource object deployment for cluster")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/deployment.go:67]:
```suggestion
				Error(err, "Failed to get status of cluster object")
```

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 14]
```diff
@@ -74,11 +74,8 @@ func (receiver *DeploymentPlugin) AggregateStatues(
 
 		deployStatus := &appsv1.DeploymentStatus{}
 		if err = util.ConvertViaJson(status, deployStatus); err != nil {
-			klog.Errorf(
-				"Failed to convert the status of cluster resource object deployment %q for cluster %q",
-				sourceObject.GetName(),
-				clusterName,
-			)
+			logger.WithValues("cluster-name", clusterName).
+				Error(err, "Failed to convert the status of cluster resource object deployment for cluster")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/deployment.go:78]:
```suggestion
				Error(err, "Failed to convert the status of cluster object")
```

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 15]
```diff
@@ -100,39 +97,28 @@ func (receiver *DeploymentPlugin) AggregateStatues(
...
 		}
 	}
 
 	newStatus, err := util.GetUnstructuredStatus(aggregatedStatus)
 	if err != nil {
-		klog.Errorf(
-			"Failed to convert the aggregatedStatus of cluster resource object deployment %q to status",
-			sourceObject.GetName(),
-		)
+		logger.Error(err, "Failed to convert the aggregatedStatus of cluster resource object deployment to status")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/deployment.go:108]:
```suggestion
		logger.Error(err, "Failed to convert aggregated status to unstructured")
```

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 16]
```diff
@@ -100,39 +97,28 @@ func (receiver *DeploymentPlugin) AggregateStatues(
...
 	}
 
 	oldStatus, _, err := unstructured.NestedMap(sourceObject.Object, common.StatusField)
 	if err != nil {
-		klog.Errorf(
-			"Failed to get old status of cluster resource object deployment %q with err: %s",
-			sourceObject.GetName(),
-			err,
-		)
+		logger.Error(err, "Failed to get old status of cluster resource object deployment")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/deployment.go:114]:
```suggestion
		logger.Error(err, "Failed to get old status of source object")
```

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 17]
```diff
@@ -100,39 +97,28 @@ func (receiver *DeploymentPlugin) AggregateStatues(
...
 
 	// update status of source object if needed
 	if !reflect.DeepEqual(newStatus, oldStatus) {
 		if err := unstructured.SetNestedMap(sourceObject.Object, newStatus, common.StatusField); err != nil {
-			klog.Errorf(
-				"Failed to set the new status of cluster resource object deployment %q with err %s",
-				sourceObject.GetName(),
-				err,
-			)
+			logger.Error(err, "Failed to set the new status of cluster resource object deployment")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/deployment.go:121]:
```suggestion
			logger.Error(err, "Failed to set the new status on source object")
```

File: pkg/controllers/statusaggregator/plugins/single_cluster_plugin.go [hunk 18]
```diff
@@ -33,9 +34,12 @@ func NewSingleClusterPlugin() *SingleClusterPlugin {
 type SingleClusterPlugin struct{}
 
 func (receiver *SingleClusterPlugin) AggregateStatues(
+	ctx context.Context,
 	sourceObject, fedObject *unstructured.Unstructured,
 	clusterObjs map[string]interface{},
 ) (*unstructured.Unstructured, bool, error) {
+	logger := klog.FromContext(ctx).WithValues("source-object-name", sourceObject.GetName())
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/single_cluster_plugin.go:41]:
```suggestion
	logger := klog.FromContext(ctx).WithValues("source-object-name", sourceObject.GetName(), "status-aggregator-plugin", "single-cluster")
```

File: pkg/controllers/statusaggregator/plugins/single_cluster_plugin.go [hunk 19]
```diff
@@ -44,11 +48,8 @@ func (receiver *SingleClusterPlugin) AggregateStatues(
 	}
 
 	if len(clusterObjs) > 1 {
-		klog.Warningf(
-			"fedObject %s associated with %d cluster objects, only 1 is supported by default status aggregator plugin.",
-			sourceObject.GetName(),
-			len(clusterObjs),
-		)
+		logger.WithValues("cluster-objs-len", len(clusterObjs)).
+			Info("FedObject associated with cluster objects, only 1 is supported by default status aggregator plugin")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/single_cluster_plugin.go:52]:
```suggestion
			Info("Federated object associated with multiple cluster objects, only 1 is supported")
```

File: pkg/controllers/statusaggregator/plugins/single_cluster_plugin.go [hunk 20]
```diff
@@ -61,11 +62,8 @@ func (receiver *SingleClusterPlugin) AggregateStatues(
 
 	newStatus, found, err := unstructured.NestedMap(clusterObj.Object, common.StatusField)
 	if err != nil {
-		klog.Errorf(
-			"Failed to get status of cluster resource object %q for cluster %q",
-			sourceObject.GetName(),
-			clusterName,
-		)
+		logger.WithValues("cluster-name", clusterName).
+			Error(err, "Failed to get status of cluster resource object for cluster")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/single_cluster_plugin.go:66]:
```suggestion
			Error(err, "Failed to get status from cluster object")
```

File: pkg/controllers/statusaggregator/plugins/single_cluster_plugin.go [hunk 21]
```diff
@@ -75,18 +73,14 @@ func (receiver *SingleClusterPlugin) AggregateStatues(
 
 	oldStatus, _, err := unstructured.NestedMap(sourceObject.Object, common.StatusField)
 	if err != nil {
-		klog.Errorf("Failed to get old status of cluster resource object %q with err: %s", sourceObject.GetName(), err)
+		logger.Error(err, "Failed to get old status of cluster resource object")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/single_cluster_plugin.go:76]:
```suggestion
		logger.Error(err, "Failed to get old status of source object")
```

File: pkg/controllers/statusaggregator/plugins/single_cluster_plugin.go [hunk 22]
```diff
@@ -75,18 +73,14 @@ func (receiver *SingleClusterPlugin) AggregateStatues(
...
 
 	// update status of source object if needed
 	if !reflect.DeepEqual(newStatus, oldStatus) {
 		if err := unstructured.SetNestedMap(sourceObject.Object, newStatus, common.StatusField); err != nil {
-			klog.Errorf(
-				"Failed to set the new status of cluster resource object %q with err %s",
-				sourceObject.GetName(),
-				err,
-			)
+			logger.Error(err, "Failed to set the new status of cluster resource object")
```
Comment (limhawjia) [inline pkg/controllers/statusaggregator/plugins/single_cluster_plugin.go:83]:
```suggestion
			logger.Error(err, "Failed to set the new status for source object")
```

File: pkg/controllers/federatedcluster/clusterjoin.go [hunk 23]
```diff
@@ -133,7 +133,7 @@ func handleNotJoinedCluster(
 
 	// 3. Get or create system namespace in the cluster, this will also tell us if the cluster is unjoinable
 
-	logger.Info(fmt.Sprintf("Create or get system namespace %s in cluster", fedSystemNamespace))
+	logger.V(1).WithValues("fed-system-namespace", fedSystemNamespace).Info("Create or get system namespace in cluster")
```
Comment (gary-lgy) [inline pkg/controllers/federatedcluster/clusterjoin.go:136]: Let's log "get" and "create" separately

File: pkg/controllers/federatedcluster/clusterjoin.go [hunk 24]
```diff
@@ -225,7 +225,7 @@ func handleNotJoinedCluster(
 
 	// 5. Cluster is joined, update condition
 
-	logger.Info("Cluster joined successfully")
+	logger.V(3).Info("Cluster joined successfully")
```
Comment (gary-lgy) [inline pkg/controllers/federatedcluster/clusterjoin.go:228]:
```suggestion
	logger.V(2).Info("Cluster joined successfully")
```
This is pretty important information, and does not have the risk of spamming. I think we can log it at 2

File: pkg/controllers/federatedcluster/clusterjoin.go [hunk 25]
```diff
@@ -244,15 +244,16 @@ func getAndSaveClusterToken(
...
+	logger := klog.FromContext(ctx)
+
+	logger.V(2).Info("Creating authorized service account")
+	saTokenSecretName, err := createAuthorizedServiceAccount(ctx, clusterKubeClient, memberSystemNamespace, cluster.Name, false)
 	if err != nil {
 		return err
 	}
 
-	logger.Info("Updating cluster secret")
+	logger.V(2).Info("Updating cluster secret")
```
Comment (gary-lgy) [inline pkg/controllers/federatedcluster/clusterjoin.go:256]:
```suggestion
	logger.V(1).Info("Updating cluster secret")
```

File: pkg/controllers/federatedcluster/clusterjoin.go [hunk 26]
```diff
@@ -287,17 +288,18 @@ func createAuthorizedServiceAccount(
 	memberSystemNamespace *corev1.Namespace,
 	clusterName string,
 	errorOnExisting bool,
-	logger klog.Logger,
 ) (string, error) {
+	logger := klog.FromContext(ctx).WithValues("member-service-account-name", MemberServiceAccountName)
```
Comment (gary-lgy) [inline pkg/controllers/federatedcluster/clusterjoin.go:292]:
```suggestion
	logger := klog.FromContext(ctx).WithValues("member-service-account-name", MemberServiceAccountName)
	ctx = klog.NewContext(ctx, logger)
```
Logger with the new key value needs to be propagated in the new context.

File: pkg/controllers/federatedcluster/clusterjoin.go [hunk 27]
```diff
@@ -309,10 +311,10 @@ func createAuthorizedServiceAccount(
 	if err != nil {
 		return "", fmt.Errorf("error creating service account token secret %s : %w", MemberServiceAccountName, err)
 	}
-	logger.Info(fmt.Sprintf("Created service account token secret %s for service account %v", saTokenSecretName, MemberServiceAccountName))
+	logger.V(1).WithValues("sa-toke-secret-name", saTokenSecretName).Info("Created service account token secret for service account")
```
Comment (gary-lgy) [inline pkg/controllers/federatedcluster/clusterjoin.go:314]:
```suggestion
	logger.V(1).WithValues("sa-token-secret-name", saTokenSecretName).Info("Created service account token secret for service account")
```
Let's add this key-value pair to the logger before we create the service account token secret.

File: pkg/controllers/follower/controller.go [hunk 28]
```diff
@@ -492,7 +488,7 @@ func (c *Controller) reconcileFollower(
 		)
 		return worker.StatusError
 	} else if updated {
-		logger.V(4).Info("Updated follower to sync with leaders")
+		logger.V(1).Info("Updated follower to sync with leaders")
```
Comment (gary-lgy) [inline pkg/controllers/follower/controller.go:491]: Let's also log before updating the follower

Comment (gary-lgy) [reply pkg/controllers/follower/controller.go:491]: Let's also log before updating the leader.

File: pkg/controllers/policyrc/controller.go [hunk 29]
```diff
@@ -94,6 +99,7 @@ func newController(controllerConfig *util.ControllerConfig,
 		name:       userAgent,
 		typeConfig: typeConfig,
 		metrics:    controllerConfig.Metrics,
+		logger:     klog.LoggerWithName(klog.Background(), ControllerName),
```
Comment (gary-lgy) [inline pkg/controllers/policyrc/controller.go:102]:
```suggestion
		logger:     klog.LoggerWithValues(klog.Background(), "controller", ControllerName, "ftc", typeConfig.Name),
```

File: pkg/controllers/policyrc/controller.go [hunk 30]
```diff
@@ -76,7 +81,7 @@ func StartController(controllerConfig *util.ControllerConfig,
 	if err != nil {
 		return err
 	}
-	klog.V(4).Infof("Starting policyrc controller for %q", typeConfig.GetFederatedType().Kind)
+	controller.logger.WithValues("federated-kind", typeConfig.GetFederatedType().Kind).Info("Starting policyrc controller")
```
Comment (gary-lgy) [inline pkg/controllers/policyrc/controller.go:84]:
```suggestion
	controller.logger.Info("Starting policyrc controller")
```

File: pkg/controllers/policyrc/controller.go [hunk 31]
```diff
@@ -221,15 +230,14 @@ func (c *Controller) HasSynced() bool {
 
 func (c *Controller) reconcileCount(qualifiedName common.QualifiedName) worker.Result {
 	federatedKind := c.typeConfig.GetFederatedType().Kind
-	key := qualifiedName.String()
+	logger := c.logger.WithValues("federated-kind", federatedKind, "object", qualifiedName.String())
```
Comment (gary-lgy) [inline pkg/controllers/policyrc/controller.go:233]:
```suggestion
	logger := c.logger.WithValues("object", qualifiedName.String())
```

File: pkg/controllers/policyrc/controller.go [hunk 32]
```diff
@@ -221,15 +230,14 @@ func (c *Controller) HasSynced() bool {
...
 
 	c.metrics.Rate("policyrc-count-controller.throughput", 1)
-	klog.V(4).Infof("policyrc count controller for %v starting to reconcile %v", federatedKind, key)
+	logger.V(3).Info("Policyrc count controller starting to reconcile")
 	startTime := time.Now()
 	defer func() {
 		c.metrics.Duration("policyrc-count-controller.latency", startTime)
-		klog.V(4).
-			Infof("policyrc count controller for %v finished reconciling %v (duration: %v)", federatedKind, key, time.Since(startTime))
+		logger.V(3).WithValues("duration", time.Since(startTime)).Info("Policyrc count controller finished reconciling")
```
Comment (gary-lgy) [inline pkg/controllers/policyrc/controller.go:240]: Pls log the status as well

File: pkg/controllers/policyrc/controller.go [hunk 33]
```diff
@@ -276,17 +284,14 @@ func (c *Controller) reconcilePersist(
 	counter *Counter,
 ) worker.Result {
 	federatedKind := c.typeConfig.GetFederatedType().Kind
-	key := qualifiedName.String()
+	logger := c.logger.WithValues("metric-name", metricName, "federated-kind", federatedKind, "object", qualifiedName.String())
```
Comment (gary-lgy) [inline pkg/controllers/policyrc/controller.go:287]:
```suggestion
	logger := c.logger.WithValues("object", qualifiedName.String())
```

File: pkg/controllers/statusaggregator/controller.go [hunk 34]
```diff
@@ -243,47 +249,48 @@ func (a *StatusAggregator) HasSynced() bool {
 func (a *StatusAggregator) reconcile(qualifiedName common.QualifiedName) worker.Result {
 	sourceKind := a.typeConfig.GetSourceType().Kind
 	key := qualifiedName.String()
+	logger := a.logger.WithValues("object", key, "source-kind", sourceKind)
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/controller.go:252]:
```suggestion
	logger := a.logger.WithValues("object", key)
```

File: pkg/controllers/statusaggregator/controller.go [hunk 35]
```diff
@@ -243,47 +249,48 @@ func (a *StatusAggregator) HasSynced() bool {
...
+	ctx := klog.NewContext(context.TODO(), logger)
 
 	a.metrics.Rate("status-aggregator.throughput", 1)
-	klog.V(4).Infof("status aggregator for %v starting to reconcile %v", sourceKind, key)
+	logger.V(3).Info("Status aggregator starting to reconcile")
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/controller.go:256]:
```suggestion
	logger.V(3).Info("Starting to reconcile")
```

File: pkg/controllers/statusaggregator/controller.go [hunk 36]
```diff
@@ -243,47 +249,48 @@ func (a *StatusAggregator) HasSynced() bool {
...
 	startTime := time.Now()
 	defer func() {
 		a.metrics.Duration("status-aggregator.latency", startTime)
-		klog.V(4).
-			Infof("status aggregator for %v finished reconciling %v (duration: %v)", sourceKind, key, time.Since(startTime))
+		logger.V(3).Info("Status aggregator finished reconciling")
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/controller.go:260]:
```suggestion
		logger.V(3).Info("Finished reconciling")
```

File: pkg/controllers/statusaggregator/controller.go [hunk 37]
```diff
@@ -243,47 +249,48 @@ func (a *StatusAggregator) HasSynced() bool {
...
-		utilruntime.HandleError(err)
+		logger.Error(err, "Failed to get cluster objs")
 		return worker.StatusError
 	}
 
-	newObj, needUpdate, err := a.plugin.AggregateStatues(sourceObject.DeepCopy(), fedObject, clusterObjs)
+	newObj, needUpdate, err := a.plugin.AggregateStatues(ctx, sourceObject.DeepCopy(), fedObject, clusterObjs)
 	if err != nil {
-		utilruntime.HandleError(err)
+		logger.Error(err, "Failed to aggregate statues")
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/controller.go:293]:
```suggestion
		logger.Error(err, "Failed to aggregate statuses")
```

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 38]
```diff
@@ -74,11 +74,8 @@ func (receiver *DeploymentPlugin) AggregateStatues(
 
 		deployStatus := &appsv1.DeploymentStatus{}
 		if err = util.ConvertViaJson(status, deployStatus); err != nil {
-			klog.Errorf(
-				"Failed to convert the status of cluster resource object deployment %q for cluster %q",
-				sourceObject.GetName(),
-				clusterName,
-			)
+			logger.WithValues("cluster-name", clusterName).
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/plugins/deployment.go:77]: nit: we can create a sublogger with `cluster-name` key at the start of the for loop.

**Comment by limhawjia** [comment]
lgtm 
@gary-lgy ptal

**Review by limhawjia — APPROVED** [review]

File: pkg/controllers/statusaggregator/controller.go [hunk 39]
```diff
@@ -243,47 +248,48 @@ func (a *StatusAggregator) HasSynced() bool {
...
 
 	a.metrics.Rate("status-aggregator.throughput", 1)
-	klog.V(4).Infof("status aggregator for %v starting to reconcile %v", sourceKind, key)
+	logger.V(3).Info("Starting to reconcile")
 	startTime := time.Now()
 	defer func() {
 		a.metrics.Duration("status-aggregator.latency", startTime)
-		klog.V(4).
-			Infof("status aggregator for %v finished reconciling %v (duration: %v)", sourceKind, key, time.Since(startTime))
+		logger.V(3).Info("Finished reconciling")
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/controller.go:259]: Let's add status and duration to the log

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 40]
```diff
@@ -39,9 +39,13 @@ func NewDeploymentPlugin() *DeploymentPlugin {
 }
 
 func (receiver *DeploymentPlugin) AggregateStatues(
+	ctx context.Context,
 	sourceObject, fedObject *unstructured.Unstructured,
 	clusterObjs map[string]interface{},
 ) (*unstructured.Unstructured, bool, error) {
+	logger := klog.FromContext(ctx).WithValues("source-object-name", sourceObject.GetName(), "status-aggregator-plugin", "deployments")
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/plugins/deployment.go:46]: `source-object-name` should have been added under `object`.

File: pkg/controllers/statusaggregator/plugins/deployment.go [hunk 41]
```diff
@@ -55,16 +59,14 @@ func (receiver *DeploymentPlugin) AggregateStatues(
 	}
 
 	for clusterName, clusterObj := range clusterObjs {
+		logger := klog.FromContext(ctx).WithValues("cluster-name", clusterName)
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/plugins/deployment.go:62]:
```suggestion
		logger := logger.WithValues("cluster-name", clusterName)
```

File: pkg/controllers/statusaggregator/plugins/deployment.go (hunk 40 above)
Comment (wy-lucky) [reply pkg/controllers/statusaggregator/plugins/deployment.go]: get, object is added during logger initialization

File: pkg/controllers/statusaggregator/plugins/single_cluster_plugin.go [hunk 42]
```diff
@@ -33,9 +34,12 @@ func NewSingleClusterPlugin() *SingleClusterPlugin {
 type SingleClusterPlugin struct{}
 
 func (receiver *SingleClusterPlugin) AggregateStatues(
+	ctx context.Context,
 	sourceObject, fedObject *unstructured.Unstructured,
 	clusterObjs map[string]interface{},
 ) (*unstructured.Unstructured, bool, error) {
+	logger := klog.FromContext(ctx).WithValues("source-object-name", sourceObject.GetName(), "status-aggregator-plugin", "single-cluster")
```
Comment (gary-lgy) [inline pkg/controllers/statusaggregator/plugins/single_cluster_plugin.go:41]:
```suggestion
	logger := klog.FromContext(ctx).WithValues("status-aggregator-plugin", "single-cluster")
```

**Review by gary-lgy — APPROVED** [review]

**Pull request merged**